File: hw/yanez/u3/HW16MazeGenerator/controller/RandomMazeGenerator.py
import random
from typing import List, Optional
from model.Direction import Direction
from model.Room import Room
from controller.MazeGenerator import MazeGenerator

class RandomMazeGenerator(MazeGenerator):
    def generate(self, rooms: List[List[Room]]) -> None:
        stack = []
        start = rooms[0][0]
        start.set_visited(True)
        stack.append(start)

        while stack:
            current = stack[-1]
            unvisited_neighbors = self.get_unvisited_neighbors(current, rooms)

            if unvisited_neighbors:
                dir = random.choice(unvisited_neighbors)
                neighbor = self.get_neighbor(current, dir, rooms)

                current.get_wall(dir).open()
                neighbor.get_wall(Direction.opposite(dir)).open()

                neighbor.set_visited(True)
                stack.append(neighbor)
            else:
                stack.pop()

    def get_unvisited_neighbors(self, room: Room, rooms: List[List[Room]]) -> List[Direction]:
        neighbors = []
        for dir in Direction:
            neighbor = self.get_neighbor(room, dir, rooms)
            if neighbor is not None and not neighbor.is_visited():
                neighbors.append(dir)
        return neighbors

    def get_neighbor(self, room: Room, dir: Direction, rooms: List[List[Room]]) -> Optional[Room]:
        new_row = room.get_row() + dir.row_offset
        new_col = room.get_col() + dir.col_offset

        if 0 <= new_row < len(rooms) and 0 <= new_col < len(rooms[0]):
            return rooms[new_row][new_col]
        return None

File: hw/bonilla/U3/HW17MazeGenerator/HW16MazeGenerator/src/main/java/ec/edu/espe/hw16mazegenerator/controller/RandomMazeGenerator.java
package ec.edu.espe.hw16mazegenerator.controller;

import ec.edu.espe.hw16mazegenerator.model.Direction;
import ec.edu.espe.hw16mazegenerator.model.Room;
import java.util.*;

/**
 *
 * @author Bonilla David SoftCrafters
 */

public class RandomMazeGenerator implements MazeGenerator {

    @Override
    public void generate(Room[][] rooms) {
        Stack<Room> stack = new Stack<>();
        Room start = rooms[0][0];
        start.setVisited(true);
        stack.push(start);

        while (!stack.isEmpty()) {
            Room current = stack.peek();
            List<Direction> unvisitedNeighbors = getUnvisitedNeighbors(current, rooms);

            if (!unvisitedNeighbors.isEmpty()) {
                Direction dir = unvisitedNeighbors.get(new Random().nextInt(unvisitedNeighbors.size()));
                Room neighbor = getNeighbor(current, dir, rooms);

                current.getWall(dir).open();
                neighbor.getWall(Direction.opposite(dir)).open();

                neighbor.setVisited(true);
                stack.push(neighbor);
            } else {
                stack.pop();
            }
        }
    }

    private List<Direction> getUnvisitedNeighbors(Room room, Room[][] rooms) {
        List<Direction> neighbors = new ArrayList<>();

        for (Direction dir : Direction.values()) {
            Room neighbor = getNeighbor(room, dir, rooms);
            if (neighbor != null && !neighbor.isVisited()) {
                neighbors.add(dir);
            }
        }
        return neighbors;
    }

    private Room getNeighbor(Room room, Direction dir, Room[][] rooms) {
        int newRow = room.getRow() + dir.getRowOffset();
        int newCol = room.getCol() + dir.getColOffset();

        if (newRow >= 0 && newRow < rooms.length && newCol >= 0 && newCol < rooms[0].length) {
            return rooms[newRow][newCol];
        }
        return null;
    }
}
